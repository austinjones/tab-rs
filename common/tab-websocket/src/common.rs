use crate::WebSocket;
use futures::SinkExt;
use log::error;
use serde::Serialize;
use tungstenite::{Error, Message};

pub async fn send_message<Msg: Serialize>(websocket: &mut WebSocket, message: Msg) {
    let encoded = bincode::serialize(&message);

    if let Err(e) = encoded {
        error!("failed to encode message: {}", e);
        return;
    }

    let sent = websocket.send(Message::Binary(encoded.unwrap())).await;
    if let Err(e) = sent {
        error!("failed to send message: {}", e);
    }
}

pub async fn send_close(websocket: &mut WebSocket) {
    if let Err(e) = websocket.send(Message::Close(None)).await {
        match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                return;
            }
            _ => {
                error!("failed to send close frame: {}", e);
            }
        }
    }
}
