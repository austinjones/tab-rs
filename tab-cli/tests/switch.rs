use insta::assert_snapshot;

mod common;
use common::*;

// #[tokio::test]
async fn switch() -> anyhow::Result<()> {
    let mut session = TestSession::new()?;

    let result = session
        .command()
        .tab("switch/a/")
        .delay_ms(1000)
        .stdin("echo a\n")
        .delay_ms(200)
        .stdin_bytes(&[23u8])
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("create_a", result.stdout);

    let result = session
        .command()
        .tab("switch/b/")
        .delay_ms(1000)
        .stdin("echo b\n")
        .delay_ms(200)
        .stdin_bytes(&[23u8])
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("create_b", result.stdout);

    let result = session
        .command()
        .tab("switch/a/")
        .delay_ms(1000)
        .stdin("$TAB_BIN switch/b/\n")
        .delay_ms(200)
        .stdin("exit\n")
        .run()
        .await?;

    assert_eq!(Some(0), result.exit_status.code());
    assert_snapshot!("after", result.stdout);

    Ok(())
}
