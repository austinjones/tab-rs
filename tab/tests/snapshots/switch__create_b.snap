---
source: tab-cli/tests/switch.rs
expression: result.stdout
---
$ echo b
b
$ 
